src/Storage/Device/Local.php

```diff
@@ -48,67 +48,146 @@ public function getRoot(): string
 
     /**
      * @param string $filename
+     * @param string $prefix
      *
      * @return string
      */
-    public function getPath($filename): string
+    public function getPath(string $filename, string $prefix = null): string
     {
         $path = '';
 
         for ($i = 0; $i < 4; ++$i) {
             $path = ($i < \strlen($filename)) ? $path . DIRECTORY_SEPARATOR . $filename[$i] : $path . DIRECTORY_SEPARATOR . 'x';
         }
-
+        if(!is_null($prefix)) {
+            $path = $prefix . DIRECTORY_SEPARATOR . $path;
+        }
         return $this->getRoot() . $path . DIRECTORY_SEPARATOR . $filename;
     }
 
     /**
      * Upload.
      *
      * Upload a file to desired destination in the selected disk.
+     * return number of chunks uploaded or 0 if it fails.
      *
      * @param string $source
      * @param string $path
+     * @param int $chunk
+     * @param int $chunks
+     * @param array $metadata
      *
      * @throws \Exception
      *
-     * @return bool
+     * @return int
      */
-    public function upload($source, $path): bool
+    public function upload(string $source, string $path, int $chunk = 1, int $chunks = 1, array &$metadata = []): int
```

Comment:
What is metadata used for? why are we using a reference? maybe we can document it?

Reply:
We discussed this, as S3 chunked upload requires some other information like upload id and each chunks ID, we agreed to use this reference object. This is not required for local. Where do you think would be a good place to document this?